Support for port groups in reservation logic

anjan-keysight opened the pull request:
@greg-dennis I have synced with the recent changes in main and made necessary changes for the port group support.

greg-dennis commented on knebind/solver/solver.go:
```diff
@@ -89,15 +88,65 @@ func Solve(tb *opb.Testbed, topo *tpb.Topology) (*binding.Reservation, error) {
+		if group == "" {
+			group = dstPort.GetGroup()
+		}
+		devLinkPtr := &devLink{srcDev: srcDev, dstDev: dstDev, srcPort: srcPort, dstPort: dstPort, group: group}
```
Rename to "dLink" for consistency with the rest of the file. Go style is typically to not include any part of the type in the variable name.

greg-dennis commented on knebind/solver/solver.go:
```diff
@@ -115,16 +164,36 @@ func Solve(tb *opb.Testbed, topo *tpb.Topology) (*binding.Reservation, error) {
+		if group == "" {
+			group = dstNode.Interfaces[l.GetZInt()].GetGroup()
+		}
+		nodeLinkPtr := &nodeLink{srcNode: srcNode, dstNode: dstNode, srcIntf: srcIntf, dstIntf: dstIntf, group: group}
```
rename to "nLink" for consistency and Go style

greg-dennis commented on knebind/solver/solver.go:
```diff
@@ -89,15 +88,65 @@ func Solve(tb *opb.Testbed, topo *tpb.Topology) (*binding.Reservation, error) {
+		}
+		devLinkPtr := &devLink{srcDev: srcDev, dstDev: dstDev, srcPort: srcPort, dstPort: dstPort, group: group}
+		if _, ok := s.dev2Links[srcDev][dstDev]; !ok {
+			pgPtr := &devPG{groupMap: make(map[string][]*devLink), sortedNames: []string{}, totalLinks: 0}
```
rename to "pg". Drop "Ptr" suffixes throughout ...

greg-dennis commented on knebind/solver/solver.go:
```diff
@@ -89,15 +88,65 @@ func Solve(tb *opb.Testbed, topo *tpb.Topology) (*binding.Reservation, error) {
+	// and go down that list; all un-assigned interfaces can be allocated to regular port.
+	// pgMap - map of port group name and membership size
+	// returns list of sorted port group names in the order of membership size, and number of total links
+	sortGroupBySize := func(pgMap map[string]int) ([]string, int) {
```
Do we have a test or benchmark that demonstrates the benefit of sorting? If not, postpone until there is one.

anjan-keysight replied:
Yes, I will add one. Without sorting, it may not 'solve' for some correct configurations.

greg-dennis commented on knebind/solver/solver.go:
```diff
@@ -115,16 +164,36 @@ func Solve(tb *opb.Testbed, topo *tpb.Topology) (*binding.Reservation, error) {
-		s.node2Links[srcNode][dstNode] = append(s.node2Links[srcNode][dstNode], nLink)
-		s.node2Links[dstNode][srcNode] = append(s.node2Links[dstNode][srcNode], nLink)
+		group := srcNode.Interfaces[l.GetAInt()].GetGroup()
+		if group == "" {
```
If both the srcNode and dstNode have groups, this would appear naively to drop information that might be necessary to arrive at a correct solution. I think there is an assumption there about the port groups that is unstated. Could you document the assumption and/or rationale that allows the name of the destination group to be ignored?

anjan-keysight replied:
Sure, will add comment here.

greg-dennis replied:
I'm still concerned about the case, both with the device links and the node links, where you have port groups at both ends and links in _both_ directions. It seems like we may lose the constraint that those link belong to the same "group". I don't see a test for this case.

anjan-keysight replied:
I will error out if both ends have group specified, it is unsupported for new. Will also add a testcase for this.

octpetre commented:
@greg-dennis could you please review?

greg-dennis commented on knebind/solver/solver.go:
```diff
@@ -115,16 +164,36 @@ func Solve(tb *opb.Testbed, topo *tpb.Topology) (*binding.Reservation, error) {
+		}
+		nodeLinkPtr := &nodeLink{srcNode: srcNode, dstNode: dstNode, srcIntf: srcIntf, dstIntf: dstIntf, group: group}
+		if _, ok := s.node2Links[srcNode][dstNode]; !ok {
+			pgPtr := &nodePG{groupMap: make(map[string][]*nodeLink), sortedNames: []string{}, totalLinks: 0}
```
no need to specify sortedNames and totalLinks. Go style is to let slices initialize to nil and integers will default to zero already.

greg-dennis commented on knebind/solver/solver.go:
```diff
@@ -115,16 +164,36 @@ func Solve(tb *opb.Testbed, topo *tpb.Topology) (*binding.Reservation, error) {
+	pNode := make(map[*tpb.Node]bool)
+	for srcNode := range s.node2Links {
+		for dstNode, nodePGPtr := range s.node2Links[srcNode] {
+			if _, ok := pNode[dstNode]; ok {
```
just "if pNode[dstNode] {"

greg-dennis commented on knebind/solver/solver.go:
```diff
@@ -336,7 +448,7 @@ func (s *solver) buildLink2Links(d2n map[*opb.Device]*tpb.Node) map[*devLink][]*
 
 func (s *solver) nodeMatches(dev *opb.Device, isATE bool) ([]*tpb.Node, error) {
-	var nodeList []*tpb.Node
+	nodeList := []*tpb.Node{}
```
revert this line. Go style is let slices initialize to nil. Nearly all code and operations on slices don't distinguish a nil slice from an empty slice

greg-dennis commented on knebind/solver/solver.go:
```diff
@@ -418,28 +528,72 @@ func (s *solver) genLinkRecurse(
+// All regular links (with no group) are added at the end.
+// m - is the map of all testbed links and corresponding potential topology links for mapping
+// returns list of key links in the desirable processing order, with all regular links added at the end
+func (s *solver) arrangeLinkKeys(m map[*devLink][]*nodeLink) []*devLink {
```
similarly, only include if a test or benchmark can distinguish between having and not having this logic

anjan-keysight replied:
Will add test for this too.

greg-dennis commented on knebind/solver/solver.go:
```diff
@@ -418,28 +528,72 @@ func (s *solver) genLinkRecurse(
+// returns list of key links in the desirable processing order, with all regular links added at the end
+func (s *solver) arrangeLinkKeys(m map[*devLink][]*nodeLink) []*devLink {
+	// All regular link (port) allocation should happen last
+	allRlinks := []*devLink{}
```
just "var allRlinks []*devLink"
let slices initialize to nil here and below

greg-dennis commented:
> @greg-dennis could you please review?

@octpetre, just saw this ... yes, my review was already underway.

anjan-keysight commented:
@greg-dennis I have incorporated the review comments, except one testcase for the arrangeLinkKeys func. We need a testcase with large number of port groups and membership size. That function tries to trim incorrect solution paths for link mapping (just like the first PR for devices). Without that, the solver will solve eventually after lot of retries.
I will start working on a benchmark test for large link groups.

anjan-keysight commented:
@greg-dennis I have fixed one slice index overflow issue in the latest checkin and also modified the TestSolveGroup to bring out the benefit of arrangeLinkKeys function. It take just over a sec to run the test, without that function it will be close to a minute.

greg-dennis commented on knebind/solver/solver_test.go:
```diff
@@ -283,6 +283,217 @@ func TestSolve(t *testing.T) {
+
+	res, err := Solve(tb, topo)
+	if err != nil {
+		t.Fatalf("Got unexpected error: %v", err)
```
Go style says to name the failing function first:
  t.Fatalf("Solve got unexpected error: %v", err)

greg-dennis commented on knebind/solver/solver_test.go:
```diff
@@ -283,6 +283,217 @@ func TestSolve(t *testing.T) {
+	if err != nil {
+		t.Fatalf("Got unexpected error: %v", err)
+	}
+	if res == nil {
```
Is there exactly one solution, and if so, could you compare to exactly that solution?

anjan-keysight replied:
There can be multiple solutions, all regular ports can be assigned any of the regular intf. So, we can not determine an exact solution here.

anjan-keysight commented on knebind/solver/solver.go:
```diff
@@ -418,28 +559,71 @@ func (s *solver) genLinkRecurse(
+				} else {
+					// This is to keep track of which iteration is adding entry to grpRes map,
+					// so as to delete it when res entry is deleted for that iteration
+					grpResUpdated = true
```
Thanks for pointing this out! There was an issue in the original condition check; rectified it.

anjan-keysight commented:
@greg-dennis I have addressed those latest comments. I am also erroring out when group is specified on both ends of a link (unsupported config), both for testbed and topology. Have added testcases for those as well.

greg-dennis commented:
This looks good. I will proceed to integrate it externally.

greg-dennis commented:
This has been exported back out after internal review.
